libraries/node-core-library/src/RealNodeModulePath.ts

```diff
@@ -0,0 +1,156 @@
+          return `${realNodeModulePathInternal(input.slice(0, nodeModulesIndex))}${input.slice(nodeModulesIndex)}`;
+        }
+
+        linkStart = linkEnd;
+        linkEnd = input.indexOf(pathSeparator, linkStart + 1);
+      }
+
+      // No trailing separator, so the link is the last path segment
+      if (linkEnd < 0) {
+        linkEnd = input.length;
+      }
+
+      const linkCandidate: string = input.slice(0, linkEnd);
+      // Check if the link is a symlink
+      const linkTarget: string | undefined = tryReadLink(linkCandidate);
+      if (linkTarget && path.isAbsolute(linkTarget)) {
+        // Absolute path, combine the link target with any remaining path segments
+        // Cache the resolution to avoid the readlink call in subsequent calls
+        cache.set(linkCandidate, linkTarget);
+        cache.set(linkTarget, linkTarget);
+        return `${linkTarget}${input.slice(linkEnd)}`;
+      }
+
+      // Relative path or does not exist
+      // Either way, the path before the last node_modules could itself be in a node_modules folder
+      // So resolve the base path to find out what paths are relative to
+      const realpathBeforeNodeModules: string = realNodeModulePathInternal(input.slice(0, nodeModulesIndex));
+      if (linkTarget) {
+        // Relative path in symbolic link. Should be resolved relative to real path of base path.
+        const resolvedTarget: string = path.resolve(
+          `${realpathBeforeNodeModules}${input.slice(nodeModulesIndex, linkStart)}`,
+          linkTarget
+        );
+        // Cache the result of the combined resolution to avoid the readlink call in subsequent calls
+        cache.set(linkCandidate, resolvedTarget);
+        cache.set(resolvedTarget, resolvedTarget);
+        return `${resolvedTarget}${input.slice(linkEnd)}`;
+      }
+
+      // No symlink, so just return the real path before the last node_modules combined with the
+      // subsequent path segments
+      return `${realpathBeforeNodeModules}${input.slice(nodeModulesIndex)}`;
+    }
+
+    this.realNodeModulePath = (input: string) => {
+      return realNodeModulePathInternal(path.normalize(input));
+    };
+  }
+
+  /**
+   * Clears the cache of resolved symlinks.
+   * @public
+   */
+  public clearCache(): void {
+    this._cache.clear();
+  }
+
+  /**
+   * Tries to read a symbolic link at the specified path.
+   * If the input is not a symbolic link, returns undefined.
+   * @param link - The link to try to read
+   * @returns The target of the symbolic link, or undefined if the input is not a symbolic link
+   */
+  private _tryReadLink(link: string): string | undefined {
+    const cached: string | undefined = this._cache.get(link);
+    if (cached) {
+      return cached;
+    }
+
+    try {
+      return this._readlinkSync(link, 'utf8');
+    } catch (err) {
+      // EISDIR and EINVAL both indicate the input is not a symbolic link
+      if (err.code !== 'EISDIR' && err.code !== 'EINVAL') {
```

Comment:
Maybe I'm misunderstanding the error code here, but what if the target file itself is a symlink, not just a file that you path to via symlinks?

Reply:
This code assumes that the referent of a symlink is never itself a symlink. This resolver is optimized specifically for an environment like a Rush repository, where the only symbolic links present are those used by pnpm.